sql/moz-fx-data-shared-prod/mozilla_org_derived/ga_sessions_v1/query.sql

```diff
@@ -0,0 +1,154 @@
+      THEN "desktop_esr"
+    WHEN target LIKE "Firefox Developer Edition%"
+      THEN "desktop_developer_edition"
+    WHEN target LIKE "Firefox Beta%"
+      THEN "desktop_beta"
+    WHEN target LIKE "Firefox Nightly Edition%"
+      THEN "desktop_nightly"
+    WHEN target LIKE "Firefox for Android%"
+      THEN "android_release"
+    WHEN target LIKE "Firefox Beta Android%"
+      THEN "android_beta"
+    WHEN target LIKE "Firefox for iOS%"
+      THEN "ios_release"
+    ELSE NULL
+  END
+);
+
+WITH daily_sessions AS (
+  SELECT
+    mozfun.ga.nullify_string(clientId) AS ga_client_id,
+    -- visitId (or sessionId in GA4) is guaranteed unique only among one client, look at visitId here https://support.google.com/analytics/answer/3437719?hl=en
+    CONCAT(mozfun.ga.nullify_string(clientId), CAST(visitId AS STRING)) AS ga_session_id,
+    MIN(PARSE_DATE('%Y%m%d', date)) AS session_date,
+    MIN(visitNumber) = 1 AS is_first_session,
+    MIN(visitNumber) AS session_number,
+    ARRAY_CONCAT_AGG(hits) AS hits,
+    SUM(totals.timeOnSite) AS time_on_site,
+    SUM(totals.pageviews) AS pageviews,
+    /* Geos */
+    MIN_BY(geoNetwork.country, visitStartTime) AS country,
+    MIN_BY(geoNetwork.region, visitStartTime) AS region,
+    MIN_BY(geoNetwork.city, visitStartTime) AS city,
+    /* Attribution */
+    MIN_BY(
+      CAST(trafficSource.adwordsClickInfo.campaignId AS STRING),
+      visitStartTime
+    ) AS campaign_id,
+    MIN_BY(trafficSource.campaign, visitStartTime) AS campaign,
+    MIN_BY(trafficSource.source, visitStartTime) AS source,
+    MIN_BY(trafficSource.medium, visitStartTime) AS medium,
+    MIN_BY(trafficSource.keyword, visitStartTime) AS term,
+    MIN_BY(trafficSource.adContent, visitStartTime) AS content,
+    ARRAY_AGG(
+      mozfun.ga.nullify_string(trafficSource.adwordsClickInfo.gclId) IGNORE NULLS
+    )[0] AS gclid,
+    /* Device */
+    MIN_BY(device.deviceCategory, visitStartTime) AS device_category,
+    MIN_BY(device.mobileDeviceModel, visitStartTime) AS mobile_device_model,
+    MIN_BY(device.mobileDeviceInfo, visitStartTime) AS mobile_device_string,
+    MIN_BY(device.operatingSystem, visitStartTime) AS os,
+    MIN_BY(device.operatingSystemVersion, visitStartTime) AS os_version,
+    MIN_BY(device.language, visitStartTime) AS language,
+    MIN_BY(device.browser, visitStartTime) AS browser,
+    MIN_BY(device.browserVersion, visitStartTime) AS browser_version,
+  FROM
+    `moz-fx-data-marketing-prod.65789850.ga_sessions_*`
+  WHERE
+    -- This table is partitioned, so we only process the data from session_date
+    -- To handle late-arriving data, we process 3 days of data each day (re-processing the past 2)
+    -- as separate Airflow tasks (or via bqetl backfill, I haven't decided yet)
+    --
+    -- Here, we need to take data from yesterday, just in case some of our sessions from today
+    -- actually started yesterday. If they did, they'll be filtered out in the HAVING clause
```

Comment:
Would it be acceptable to process data once it's "settled" 3 days later?

Reply:
That's probably waiting too long. I would prefer to have this updated daily (and it's usually a large majority of the data that arrives same-day).

Reply:
I can see the utility of being able to see the campaign immediately but I'd be worried about changing numbers and how peeps might feel if reports shift on them.